### PR TITLE
skywalking: use upstream host address as addressUsedAtClient in propagation header

### PR DESCRIPTION
**Pull request by Shikugawa**
Signed-off-by: Shikugawa <rei@tetrate.io>

Commit Message: skywalking: use remote IP address as addressUsedAtClient in propagation header
Additional Description: In Apache SkyWalking, the request pathname may be used as the span name, or the remote IP address may be set in propagation.
Risk Level: Low
Testing: Unit
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/router/upstream_request.cc
Comment (KBaichoo) [inline source/common/router/upstream_request.cc:60]: what makes  `parent.timeSource().systemTime` here interchangable with `stream_info.startTime()`?

**Review by KBaichoo — COMMENTED** [review]
//test/extensions/tracers/xray:xray_test  fails to build btw

File: source/common/router/upstream_request.cc
Comment (Shikugawa) [reply source/common/router/upstream_request.cc:60]: child span may take not only system time but also stream-related info in Apache SkyWalking.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks. This API change has big impact. So I think we need to think about it more carefully.

I am thinking if there are some others method to achieve your target.

For the `startSpan`, the request path can be obtained from `TraceContext` and upstream address is unnecessary for the EntrySpan of SkyWalking.

May be we can update `injectContext` to add new parameter to provides upstream info. The ExitSpan of SkyWalking can get path and also remote address here.
This API will be more natural, because `injectContext` is called when the tracing info is propagated upstream, and a parameter providing upstream info is more reasonable.

File: source/common/router/upstream_request.cc
Comment (Shikugawa) [reply source/common/router/upstream_request.cc:60]: > I thinks the this replace may results in unaccurate start time for all the child span from spawnChild.

Ah I missed it... We need revisit about it.

**Comment by Shikugawa** [comment]
@wbpcode @KBaichoo Fixed with your opinions. PTAL.

**Review by KBaichoo — COMMENTED** [review]
thanks, seems like there are still build failures

**Review by wbpcode — COMMENTED** [review]
It seems that there are still some issues that have not been resolved. The problem seems to be more complicated than expected. How to get the correct upstream when injectContext? It seems that only UpstreamRequest can get the target upstream correctly. 🤔

File: source/common/grpc/async_client_impl.cc
Comment (Shikugawa) [reply source/common/grpc/async_client_impl.cc:257]: This is quite complicated, and we do not necessarily want the destination IP address in an async request. `addressUsedAtClient` in SkyWalking is not a must to be an IP address. So I think we can fall back to the traditional behavior here. WDYT?

Comment (wbpcode) [reply source/common/grpc/async_client_impl.cc:257]: But now that we're going to modify this API, we should try to keep it correct in all scenarios as much as possible. But I admit that we don't need to use this one PR to solve all the problems.

Comment (Shikugawa) [reply source/common/grpc/async_client_impl.cc:257]: Here it injects authority to prop header, the behavior won't be replaced with this change only for async client. I think it will be processed as future work.

Comment (wbpcode) [reply source/common/grpc/async_client_impl.cc:257]: Get it. I will recommend to update the API to something like this:

```
  virtual void injectContext(TraceContext& trace_conext,
                             const Upstream::HostDescriptionConstSharedPtr& upstream) PURE;
```

For the async client, just keep upstream to be nullptr.

Comment (Shikugawa) [reply source/common/grpc/async_client_impl.cc:257]: much cleaner, thanks.

File: source/common/router/upstream_request.cc
Comment (wbpcode) [inline source/common/router/upstream_request.cc:496]: `onPoolReady` will provides `host` directly.

**Comment by wbpcode** [comment]
It looks like the work is almost done. 😄  There are still some problems with CI, some test cases are not fixed. 
 
 /wait

**Comment by wbpcode** [comment]
Hi, @Shikugawa, although we don't get a conclusion of #20454 , we still can complete this work first. And then we can discusss the method about how to inject peer address continuely in the #20454.

**Comment by wbpcode** [comment]
After discussion with @wu-sheng,  I will pick this up. Thanks for all your great contributions @Shikugawa.

File: README.md
Comment (wu-sheng) [inline README.md:17]: @wbpcode In default setting, once PR opened from a fork, you could submit codes there as you are a upstream maintainer :) I did this for years.

Comment (wbpcode) [reply README.md:17]: Get it.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks cc @Shikugawa. 

cc @mattklein123 could you take a look for the final pass when you have free time? Thanks.

File: README.md
Comment (mattklein123) [inline README.md:16]: revert

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Does this need a small release note? Otherwise LGTM, thanks.

/wait

**Pull request merged**
